### DISCUSSION TITLE
Simple way to do an authlogin for astro

### DISCUSSION THREAD

**Discussion by Morzaram** [pocketbase/pocketbase#1785] [Show and tell]
Hey everyone, I wanted to use PocketBase for an Astro site, and I wanted to let you all know how to do an admin login and then query through Astro since at this current time there isn't a PocketBase adapter.

So basically, the authorizePB function logs in to the admin and returns the new PocketBase client state. 

```typescript
import PocketBase from "pocketbase";

const pb = new PocketBase("https://motion-sensor-backend.fly.dev");

export const authorizePB = async () => {
  await pb.admins.authWithPassword(
    import.meta.env.PB_USERNAME,
    import.meta.env.PB_PASSWORD
  );
  return pb;
};
```

In the following lines, this is how I call it within my index page

await for the authorizePB to return and then use it like you would as if you were using the normal api docs

```typescript
import { authorizePB } from "../utils/pocket-base";

const pb = await authorizePB();

const records = await pb.collection("events").getFullList(200, {
  sort: "-created",
});
records.forEach((r: { [x: string]: any }) => console.log(r["id"]));
```

If anyone has any cleaner ideas for implementation, then let me know, but this is what I've got in the meantime!

Cheers!

**Comment by ganigeorgiev** [comment]
Just one minor note - the password verification could be "slow" and could take some computing resources, and if you call the above multiple times for each request it can hurt the performance of your PocketBase server.

I haven't got the time to experiment with Astro yet, but one minor optimization of the above could be to check for `pb.authStore.isValid` before making the authWithPassword request:

```js
export const authorizePB = async () => {
  if (!pb.authStore.isValid) {
    await pb.admins.authWithPassword(
      import.meta.env.PB_USERNAME,
      import.meta.env.PB_PASSWORD
    );
  }
  
  return pb;
};
```

Note that the above will work only if the `pb` instance is kept in memory/runtime, aka. it will not work if `/utils/pocket-base` is bootstrapped from the ground every time when a request is made as when dealing with workers for example.

Another alternative could be to pregenerate a long lived admin token and load it from some env variable (eg. `pb.authStore.save(token, null)`), but keep in mind that currently there is no support for API keys or permanent tokens and the max duration is ~2years (_you can also check this recent discussion - https://github.com/pocketbase/pocketbase/discussions/1778_).

**Reply by Morzaram** [reply]
Thanks for your response on this! Great points. I'll do some research about in memory/runtime login option and make a better version!

**Comment by Morzaram** [comment]
Hey, I ended up making a basic library for anyone interested! It works with Vite and Astro.
https://github.com/Morzaram/pocketbase-vite-serverside
